git-checkout pipeline fails with some repos

Here is output of a wolfi-dev/os 'make package/yajl' with the above yajl.yaml.
```
+ make package/yajl
getting package dir for ./yajl.yaml
For package pkgdir found dir: .
found package dir as .
getting source dir for package yajl with dir .
found source dir as --source-dir ./yajl
yamlfile is ./yajl.yaml
Building package yajl with version yajl-2.1.0-r3 from file ./yajl.yaml
make yamlfile=./yajl.yaml srcdirflag="--source-dir ./yajl" pkgname=yajl packages/x86_64/yajl-2.1.0-r3.apk
make[1]: Entering directory '/home/smoser/src/wolfi/os'
@SOURCE_DATE_EPOCH=1718226947 /home/smoser/go/bin/melange build ./yajl.yaml --repository-append /home/smoser/src/wolfi/os/packages --keyring-append local-melange.rsa.pub --signing-key local-melange.rsa --arch x86_64 --env-file build-x86_64.env --namespace wolfi --generate-index false  --pipeline-dir ./pipelines/  -k https://packages.wolfi.dev/os/wolfi-signing.rsa.pub -r https://packages.wolfi.dev/os --source-dir ./yajl --log-policy builtin:stderr,packages/x86_64/buildlogs/yajl-2.1.0-r3.log
2024/06/12 22:15:57 INFO melange is building:
2024/06/12 22:15:57 INFO   configuration file: ./yajl.yaml
2024/06/12 22:15:57 INFO   workspace dir: /tmp/melange-workspace-1736824283
2024/06/12 22:15:57 INFO evaluating pipelines for package requirements
2024/06/12 22:15:57 INFO   adding packages [git] for pipeline "Check out sources from git"
2024/06/12 22:15:57 INFO populating workspace /tmp/melange-workspace-1736824283 from ./yajl
2024/06/12 22:15:57 INFO --cache-dir ./melange-cache/ not a dir; skipping
2024/06/12 22:15:57 INFO building workspace in '/tmp/melange-guest-1036530814' with apko
2024/06/12 22:15:57 INFO image configuration:
2024/06/12 22:15:57 INFO   contents:
2024/06/12 22:15:57 INFO     repositories: []
2024/06/12 22:15:57 INFO     keyring:      []
2024/06/12 22:15:57 INFO     packages:     [busybox git]
2024/06/12 22:15:57 INFO   accounts:
2024/06/12 22:15:57 INFO     runas:  
2024/06/12 22:15:57 INFO     users:
2024/06/12 22:15:57 INFO       - uid=1000(build) gid=1000
2024/06/12 22:15:57 INFO     groups:
2024/06/12 22:15:57 INFO       - gid=1000(build) members=[build]
2024/06/12 22:15:58 INFO installing ca-certificates-bundle (20240315-r3)
2024/06/12 22:15:58 INFO installing wolfi-baselayout (20230201-r11)
2024/06/12 22:15:58 INFO installing ld-linux (2.39-r6)
2024/06/12 22:15:58 INFO installing glibc-locale-posix (2.39-r6)
2024/06/12 22:15:58 INFO installing glibc (2.39-r6)
2024/06/12 22:15:58 INFO installing libxcrypt (4.4.36-r7)
2024/06/12 22:15:58 INFO installing libcrypt1 (2.39-r6)
2024/06/12 22:15:58 INFO installing busybox (1.36.1-r10)
2024/06/12 22:15:58 INFO installing libexpat1 (2.6.2-r1)
2024/06/12 22:15:58 INFO installing libpcre2-8-0 (10.44-r0)
2024/06/12 22:15:58 INFO installing zlib (1.3.1-r2)
2024/06/12 22:15:58 INFO installing libunistring (1.2-r2)
2024/06/12 22:15:58 INFO installing libidn2 (2.3.7-r2)
2024/06/12 22:15:58 INFO installing libpsl (0.21.5-r2)
2024/06/12 22:15:58 INFO installing libbrotlicommon1 (1.1.0-r2)
2024/06/12 22:15:58 INFO installing libbrotlidec1 (1.1.0-r2)
2024/06/12 22:15:58 INFO installing libnghttp2-14 (1.62.1-r0)
2024/06/12 22:15:58 INFO installing libcrypto3 (3.3.0-r9)
2024/06/12 22:15:58 INFO installing libssl3 (3.3.0-r9)
2024/06/12 22:15:58 INFO installing libcurl-openssl4 (8.8.0-r0)
2024/06/12 22:15:58 INFO installing git (2.45.2-r0)
2024/06/12 22:15:58 INFO built image layer tarball as /tmp/apko-temp-2993088703/apko-x86_64.tar.gz
2024/06/12 22:15:58 INFO using /tmp/apko-temp-2993088703/apko-x86_64.tar.gz for image layer
2024/06/12 22:15:59 INFO ImgRef = /tmp/melange-guest-2285279181
2024/06/12 22:15:59 INFO running step "git-checkout"
2024/06/12 22:15:59 INFO running step "Check out sources from git"
2024/06/12 22:15:59 WARN Cloning into '/tmp/tmp.SkyGAz'...
2024/06/12 22:15:59 WARN fatal: Needed a single revision
2024/06/12 22:15:59 ERRO ERROR: failed to build package. the build environment has been preserved:
2024/06/12 22:15:59 INFO   workspace dir: /tmp/melange-workspace-1736824283
2024/06/12 22:15:59 INFO   guest dir: /tmp/melange-guest-1036530814
2024/06/12 22:15:59 INFO error during command execution: failed to build package: unable to run pipeline: exit status 128
make[1]: *** [Makefile:150: packages/x86_64/yajl-2.1.0-r3.apk] Error 1
make[1]: Leaving directory '/home/smoser/src/wolfi/os'
make: *** [Makefile:140: package/yajl] Error 2

```

--- a/e2e-tests/test-git-checkout.yaml
+++ b/e2e-tests/test-git-checkout.yaml
@@ -8,11 +8,9 @@ environment:
     packages:
       - busybox
       - git
-      - git-daemon
 vars:
   workd: /tmp/test-git-checkout-workd
   giturl: "file:///tmp/test-git-checkout-workd/repos/my-repo"
-  gitremoteurl: "git://127.6.14.21:61421/my-repo"
 pipeline:
   - name: "Create the bogus package content"
     runs: |
@@ -25,29 +23,9 @@ pipeline:
 
       ./create-git-repo "$repo"
       touch "$repo/git-daemon-export-ok"
-  - name: "Run the git daemon"
-    runs: |
-      url=${{vars.gitremoteurl}}
-      addr=${url#git://}
-      addr=${addr%%:*}
-      port=${url#git://*:}
-      port=${port%%/*}
-      set -- git daemon --listen="$addr" --port="$port" \
-          --detach --pid-file=${{vars.workd}}/pid \
-          --log-destination=stderr \
-          --base-path="${{vars.workd}}/repos/" "${{vars.workd}}/repos/"
-      echo "running daemon:" "$@"
-      "$@"
-      read pid < ${{vars.workd}}/pid
-      echo "pid is $pid"
-
-      vr() { echo "$@" 1>&2; "$@"; }
-      vr git clone "${{vars.giturl}}" my.git
-      vr git clone "${{vars.giturl}}" my2.git
   - runs: |
       vr() { echo "$@" 1>&2; "$@"; }
       vr git clone "${{vars.giturl}}" my3.git
-
   - name: "standard tag on branch"
     uses: git-checkout
     working-directory: standard

--- a/e2e-tests/test-git-checkout.yaml
+++ b/e2e-tests/test-git-checkout.yaml
@@ -1,0 +1,62 @@
+package:
+  name: test-git-checkout
+  version: 6.14
+  epoch: 0
+  description: This package mainly just tests git-checkout pipeline
+environment:
+  contents:
+    packages:
+      - busybox
+      - git
+      - git-daemon
+vars:
+  workd: /tmp/test-git-checkout-workd
+  giturl: "file:///tmp/test-git-checkout-workd/repos/my-repo"
+  gitremoteurl: "git://127.6.14.21:61421/my-repo"
+pipeline:
+  - name: "Create the bogus package content"
+    runs: |
+      echo "package does not do anything" > "${{targets.contextdir}}/README"
+  - name: "Create a git repo"
+    runs: |
+      rm -Rf ${{vars.workd}}
+      mkdir -p ${{vars.workd}}/repos
+      repo=${{vars.workd}}/repos/my-repo
+
+      ./create-git-repo "$repo"
+      touch "$repo/git-daemon-export-ok"
+  - name: "Run the git daemon"
+    runs: |
+      url=${{vars.gitremoteurl}}
+      addr=${url#git://}
+      addr=${addr%%:*}
+      port=${url#git://*:}
+      port=${port%%/*}
+      set -- git daemon --listen="$addr" --port="$port" \
+          --detach --pid-file=${{vars.workd}}/pid \
+          --log-destination=stderr \
+          --base-path="${{vars.workd}}/repos/" "${{vars.workd}}/repos/"
+      echo "running daemon:" "$@"
+      "$@"
+      read pid < ${{vars.workd}}/pid
+      echo "pid is $pid"
+
+      vr() { echo "$@" 1>&2; "$@"; }
+      vr git clone "${{vars.giturl}}" my.git
+      vr git clone "${{vars.giturl}}" my2.git
+  - runs: |
+      vr() { echo "$@" 1>&2; "$@"; }
+      vr git clone "${{vars.giturl}}" my3.git
+
+  - name: "standard tag on branch"
+    uses: git-checkout
+    working-directory: standard
+    with:
+      repository: ${{vars.giturl}}
+      tag: 2.0
+      expected-commit: 3dfc3dd573b814be48c07f7f8ae3c19a23b69865
+  - name: "check standard tag on branch"
+    working-directory: standard
+    runs: |
+      hash=$(git rev-parse --verify HEAD)
+      [ "$hash" = 3dfc3dd573b814be48c07f7f8ae3c19a23b69865 ]

--- a/e2e-tests/test-git-checkout.yaml
+++ b/e2e-tests/test-git-checkout.yaml
@@ -38,3 +38,117 @@ pipeline:
     runs: |
       hash=$(git rev-parse --verify HEAD)
       [ "$hash" = 3dfc3dd573b814be48c07f7f8ae3c19a23b69865 ]
+      cd ..
+      rm -Rf standard
+  - name: "standard no-working-directory"
+    uses: git-checkout
+    with:
+      repository: ${{vars.giturl}}
+      tag: 2.0
+      expected-commit: 3dfc3dd573b814be48c07f7f8ae3c19a23b69865
+  - name: "check standard no-working-directory"
+    runs: |
+      hash=$(git rev-parse --verify HEAD)
+      [ "$hash" = 3dfc3dd573b814be48c07f7f8ae3c19a23b69865 ]
+      [ -f create-git-repo ] ||
+         { echo "create-git-repo did not exist"; exit 1; }
+  - name: "destination"
+    uses: git-checkout
+    working-directory: destination-base
+    with:
+      repository: ${{vars.giturl}}
+      tag: 2.0
+      expected-commit: 3dfc3dd573b814be48c07f7f8ae3c19a23b69865
+      destination: dest
+  - name: "check destination"
+    working-directory: destination-base
+    runs: |
+      cd dest
+      hash=$(git rev-parse --verify HEAD)
+      [ "$hash" = 3dfc3dd573b814be48c07f7f8ae3c19a23b69865 ]
+      cd ../..
+      rm -R destination-base
+  - name: "depth 1"
+    uses: git-checkout
+    working-directory: depth-1
+    with:
+      repository: ${{vars.giturl}}
+      tag: 2.0
+      expected-commit: 3dfc3dd573b814be48c07f7f8ae3c19a23b69865
+  - name: "check depth 1"
+    working-directory: depth-1
+    runs: |
+      if out=$(git show HEAD^ 2>&1); then
+          echo "git show HEAD^ should not succeed on checkout with depth 1"
+          exit 1
+      fi
+      hash=$(git rev-parse --verify HEAD)
+      [ "$hash" = 3dfc3dd573b814be48c07f7f8ae3c19a23b69865 ]
+      cd ..
+      rm -R depth-1
+  - name: "depth -1"
+    working-directory: depth-neg-1
+    uses: git-checkout
+    with:
+      depth: -1
+      repository: ${{vars.giturl}}
+      tag: 2.0
+      expected-commit: 3dfc3dd573b814be48c07f7f8ae3c19a23b69865
+  - name: "check depth -1"
+    working-directory: depth-neg-1
+    runs: |
+      out=$(git show HEAD^ 2>&1) || {
+          echo "git show HEAD^ failed on depth -1 checkout"
+          echo "it should pass"
+          exit 1
+      }
+      hash=$(git rev-parse --verify HEAD)
+      [ "$hash" = 3dfc3dd573b814be48c07f7f8ae3c19a23b69865 ]
+      cd ..
+      rm -R depth-neg-1
+  - name: "branch without expected"
+    working-directory: branch-no-expected
+    uses: git-checkout
+    with:
+      repository: ${{vars.giturl}}
+      branch: 1.x
+  - name: "check branch without expected"
+    working-directory: branch-no-expected
+    runs: |
+      hash=$(git rev-parse --verify HEAD)
+      [ "$hash" = 4b1593d8d8038f8c0ce1e2c608c9dd89066a2a0f ]
+      cd ..
+      rm -R branch-no-expected
+  # for an annotated tag you can point to either the commit
+  # or the tag object hash object
+  - name: "annotated hash"
+    uses: git-checkout
+    working-directory: annotated-hash
+    with:
+      repository: ${{vars.giturl}}
+      tag: 2.0-annotated
+      expected-commit: 4ce5bdbf45a68a166d931dd1247878829b5c0113
+  - name: "check annotated hash"
+    working-directory: annotated-hash
+    runs: |
+      git show-ref --tags -d
+      hash=$(git rev-parse --verify HEAD)
+      [ "$hash" = 3dfc3dd573b814be48c07f7f8ae3c19a23b69865 ]
+      cd ..
+      rm -R annotated-hash
+  # special case with clone --branch if there is a tag and a branch
+  # with the same name.
+  - name: "tag and branch same name"
+    uses: git-checkout
+    working-directory: tag-and-branch
+    with:
+      repository: ${{vars.giturl}}
+      tag: dev
+      expected-commit: 2b9bb894348794bc840a2ee7553d54a1c80b9278
+  - name: "check tag and branch same name"
+    working-directory: tag-and-branch
+    runs: |
+      hash=$(git rev-parse --verify HEAD)
+      [ "$hash" = 2b9bb894348794bc840a2ee7553d54a1c80b9278 ]
+      cd ..
+      rm -R tag-and-branch

--- a/e2e-tests/test-git-checkout/create-git-repo
+++ b/e2e-tests/test-git-checkout/create-git-repo
@@ -1,0 +1,101 @@
+#!/bin/sh
+# create a repository with known hashes
+set -e
+MYDATE="Fri, 14 Jun 2024 15:21:36 -0000"
+TEMP_D=""
+
+fail() { echo "$@" 1>&2; exit 1; }
+
+wfile() {
+    name="$1" content="$2"
+    rm -f "$name"
+    printf "%s\n" "$content" > "$name"
+}
+
+gcommit() {
+    GIT_COMMITTER_DATE="$MYDATE" git commit \
+        --date="$MYDATE" "$@"
+}
+
+gtag() {
+    GIT_AUTHOR_DATE="$MYDATE" \
+        GIT_COMMITTER_DATE="$MYDATE" git tag "$@"
+}
+
+v() {
+    echo "execute:" "$@"
+    "$@"
+}
+
+cleanup() { rm -Rf "$TEMP_D"; TEMP_D=""; }
+
+TEMP_D=$(mktemp -d)
+trap cleanup EXIT
+
+dest="$1"
+if [ -e "$dest" ]; then
+    fail "do not give existing dir for dest"
+fi
+startd="$PWD"
+
+cd "$TEMP_D"
+export HOME="$PWD"
+workdir="$PWD/my-repo"
+export LC_ALL=C.UTF-8
+
+git config --global user.name "Melange Test"
+git config --global user.email meltest@example.com
+
+mkdir "$workdir"
+cd "$workdir"
+
+v git init --initial-branch=main .
+v wfile README "hello world"
+v git add README
+v gcommit -m "first commit" README
+v wfile README "data2"
+v gcommit -m "second commit" README
+
+v git checkout -b 1.x main
+v wfile README "1.0-release content"
+v gcommit -m "release 1.0" README
+v gtag 1.0 HEAD
+v gtag --annotate --message="Release 1.0" 1.0-annotated HEAD
+
+v wfile README "1.1-release content"
+v gcommit -m "release 1.1" README
+v gtag 1.1 HEAD
+
+v wfile README "1.2-staged content"
+v gcommit -m "stage 1.2 content" README
+v gtag 1.2
+v gtag --annotate --message="Release 1.2" 1.2-annotated HEAD
+
+v git checkout -b dev main
+v wfile README "stuff on dev"
+v gcommit -m "dev branch stuff" README
+# tag with same name as branch
+# https://github.com/chainguard-dev/melange/issues/1272
+v gtag dev HEAD
+v wfile README "more stuff on dev"
+v gcommit -m "more stuff on dev" README
+
+v git checkout main
+v wfile README "mainline stuff"
+v gcommit -m "mainline stuff" README
+
+v git checkout main
+v wfile README "mainline stuff 2"
+v gcommit -m "mainline stuff 2" README
+
+v git checkout -b 2.x main^
+v wfile README "2.0-release content"
+v gcommit -m "release 2.0" README
+v gtag 2.0
+v gtag --annotate --message="Release 2.0" 2.0-annotated HEAD
+
+v git checkout main
+
+v cd "$startd"
+v mv "$workdir/.git" "$dest"
+echo "wrote git dir to $dest"

--- a/pkg/build/pipelines/git-checkout.yaml
+++ b/pkg/build/pipelines/git-checkout.yaml
@@ -39,65 +39,110 @@ inputs:
 
 pipeline:
   - runs: |
-      if [ -z "${{inputs.branch}}" ] && [ -z "${{inputs.tag}}" ]; then
-        echo "Warning (git-checkout): you have not specified a branch or tag."
-      fi
+      #!/bin/sh
+      set -e
 
-      git_clone_flags=""
-      if [ "${{inputs.recurse-submodules}}" == "true" ]; then
-        git_clone_flags="--recurse-submodules"
-      fi
+      msg() { echo "[git checkout]" "$@" 1>&2; }
+      fail() { msg FAIL "$@"; exit 1; }
+      vr() { msg "execute:" "$@"; "$@"; }
 
-      [ -n '${{inputs.branch}}' ] && clone_target='--branch ${{inputs.branch}}'
-      [ -n '${{inputs.tag}}' ] && clone_target='--branch ${{inputs.tag}}'
+      main() {
+          local repo=$1 dest=${2:-.} depth=${3:-"-1"} branch=$4
+          local tag=$5 expcommit=$6 recurse=${7:-false}
+          msg "repo='$repo' dest='$dest' depth='$depth' branch='$branch'" \
+              "tag='$tag' expcommit='$expcommit' recurse='$recurse'"
 
-      workdir=$(mktemp -d)
-      mkdir -p '${{inputs.destination}}'
-      clone_fullpath=$(realpath '${{inputs.destination}}')
+          case "$recurse" in
+              true|false) :;;
+              *) fail "recurse must be true or false, not '$recurse'"
+          esac
 
-      git config --global --add safe.directory $workdir
-      git config --global --add safe.directory $clone_fullpath
-      if [ "${{inputs.depth}}" == "-1" ]; then
-        git clone $git_clone_flags $clone_target '${{inputs.repository}}' $workdir
-      else
-        git clone $git_clone_flags $clone_target --depth '${{inputs.depth}}' '${{inputs.repository}}' $workdir
-      fi
+          [ -n "$repo" ] || fail "repository not provided"
 
-      cd $workdir
-      tar -c . | (cd $clone_fullpath && tar -x)
-      rm -rf $workdir
-      cd $clone_fullpath
-      git config --global --add safe.directory $clone_fullpath
+          if [ -z "$branch" ] && [ -z "$tag" ]; then
+              msg "Warning: you have not specified a branch or tag."
+          fi
 
-      if [ -z "${{inputs.expected-commit}}" ]; then
-        echo "Warning (git-checkout): no expected-commit"
-      elif [ -n '${{inputs.branch}}' ]; then
-        git config --global advice.detachedHead false
-        remote_commit=$(git rev-parse --verify --end-of-options "refs/heads/${{inputs.branch}}")
-        if [[ '${{inputs.expected-commit}}' != "$remote_commit" ]]; then
-          echo "Error (git-checkout): expect commit ${{inputs.expected-commit}}, got $remote_commit"
-          exit 1
-        fi
-      elif [ -n '${{inputs.tag}}' ]; then
-        # If it's a tag, then it could be a lightweight or annotated tag.
-        # Lightweight tags point directly to the commit and do not have any messages, signatures, or other data.
-        # Annotated tags point to its own git object containing the tag data, with a reference to the underlying commit.
-        # We expect most tags to be using annotated tags.
+          [ -n "$expcommit" ] ||
+              msg "Warning: no expected-commit"
 
-        # Compare direct tag value
-        remote_commit=$(git rev-parse --verify --end-of-options "refs/tags/${{inputs.tag}}")
-        if [[ '${{inputs.expected-commit}}' == "$remote_commit" ]]; then
-          exit 0
-        fi
+          local flags="" depthflag="" dest_fullpath="" workdir=""
+          local remote="origin" rcfile="" rc=""
+          [ "$recurse" = "true" ] && flags="$flags --recurse-submodules"
+          [ -n "$branch" ] && flags="--branch=$branch"
+          [ -n "$tag" ] && flags="--branch=$tag"
 
-        # Try to unpeel the tag and compare the underlying value.
-        echo "Warning (git-checkout): expected commit ${{inputs.expected-commit}}, does not match tag ${remote_commit}. Attempting to unpeel tag."
+          [ "$depth" = "-1" ] || depthflag="--depth=$depth"
 
-        unpeeled_commit=$(git rev-parse --verify --end-of-options "refs/tags/${{inputs.tag}}^{}")
-        if [[ '${{inputs.expected-commit}}' != "${unpeeled_commit}" ]]; then
-          echo "Error (git-checkout): expect commit ${{inputs.expected-commit}}, got ${unpeeled_commit}"
-          exit 1
-        fi
-      else
-        echo "Error (git-checkout): no branch or tag provided"
-      fi
+          workdir=$(mktemp -d)
+          rcfile=$(mktemp)
+          mkdir -p "$dest"
+          dest_fullpath=$(realpath "$dest")
+
+          vr git config --global --add safe.directory "$workdir"
+          vr git config --global --add safe.directory "$dest_fullpath"
+
+          vr git clone "--origin=$remote" $flags \
+              ${depthflag:+"$depthflag"} "$repo" "$workdir"
+
+          vr cd "$workdir"
+          msg "tar -c . | tar -C \"$dest_fullpath\" -x"
+          ( tar -c . ; echo $? > "$rcfile") | tar -C "$dest_fullpath" -x
+          read rc < "$rcfile" || fail "failed to read rc file"
+          [ $rc -eq 0 ] || fail "tar creation in $workdir failed"
+
+          rm -rf "$workdir"
+          vr cd "$dest_fullpath"
+          vr git config --global --add safe.directory "$dest_fullpath"
+
+          local foundcommit="" tagobj=""
+          vr git config --global advice.detachedHead false
+          if [ -z "$tag" ]; then
+              foundcommit=$(git rev-parse --verify HEAD)
+              if [ -n "$expcommit" ] && [ "$expcommit" != "$foundcommit" ]; then
+                  fail "expected commit $expcommit on ${branch:-HEAD}," \
+                      " got $foundcommit"
+              fi
+              msg "tip of ${branch:-HEAD} is commit $foundcommit"
+              return 0
+          fi
+
+          # git clone --branch=X will pick the branch X if there
+          # exists both a tag and a branch by that name.
+          # since a tag was given, we want the tag.
+          vr git fetch $remote ${depthflag:+"$depthflag"} --no-tags \
+              "+refs/tags/$tag:refs/$remote/tags/$tag"
+          vr git checkout "$remote/tags/$tag"
+
+          foundcommit=$(git rev-parse --verify HEAD)
+          if [ -z "$expcommit" ] || [ "$expcommit" = "$foundcommit" ]; then
+              msg "tag $tag is $foundcommit"
+              return 0
+          fi
+
+          # If it's a tag, then it could be a lightweight or annotated tag.
+          # Lightweight tags point directly to the commit and do not have
+          # any messages, signatures, or other data.  Annotated tags point
+          # to its own git object containing the tag data, with a reference
+          # to the underlying commit.  We expect most tags to be using
+          # annotated tags.
+          tagobj=$(git rev-parse --verify --end-of-options \
+              "refs/$remote/tags/$tag")
+          if [ "$expcommit" != "$tagobj" ]; then
+              [ "$tagobj" != "$expcommit" ] &&
+                  msg "tag object hash was $tagobj"
+              fail "Expected commit $expcommit for $tag, found $foundcommit"
+          fi
+
+          msg "Warning: The provided expected-commit ($expcommit)"
+          msg "was the hash of the annotated tag object for $tag."
+          msg "Update to set expected-commit to $foundcommit"
+
+          return 0
+      }
+
+      main \
+          "${{inputs.repository}}" "${{inputs.destination}}" \
+          "${{inputs.depth}}" "${{inputs.branch}}" \
+          "${{inputs.tag}}" "${{inputs.expected-commit}}" \
+          "${{inputs.recurse-submodules}}"

